3단계 - 질문 삭제하기 리팩터링 리뷰요청

jung635 opened the pull request:
안녕하세요.

3단계 리뷰 요청 드립니다.

추가적으로 아래와 같이 몇가지 질문드립니다.

1. Question에서 answer delete까지 처리해주는 것이 맞을지 고민하였습니다. 하지만 양방향으로 맵핑하더라도 Qustion에서는 Answer를 조회만 할 수 있기 때문에 두가지 Question, Answer delete를 구분하여 Service에서 진행하였습니다. 이부분에 대해 이렇게 진행하여도 괜찮은지 의견 부탁드립니다.

2. Question 양방향 맵핑 테스트를 해보려고 getAnswerList 테스트 코드를 작성하였으나 실패하였습니다. Question의 Answer list가 항상 길이가 0으로 나오고 있습니다. 혹시 어느부분이 문제인지 알 수 있을까요?

바쁜 시간 내어 리뷰해주셔서 감사합니다. :)

jung635 committed:
docs($step3) update readme for step3

jung635 committed:
refactor($step3): refactor Answer

refactor delete method

jung635 committed:
test($step3): add test cases to Answer

jung635 committed:
feat($step3): implement Answer for added test cases

jung635 committed:
test($step3): add test cases to Question

jung635 committed:
feat($step3): implement Question for added test cases

jung635 committed:
refactor($step3): refactor QuestionRepositoryTest

refactor variable name

jung635 committed:
refacotr($step3): refactor Answer and Question

delete the fucntion that alreade existed

jung635 committed:
feat($step3): change User

delete equals

jung635 committed:
feat($stetp3): add Answers and DeleteHistories

move delete logic in QnaService to Answers and Question
and add Answers and DeleteHistories

jung635 committed:
refactor($step3): refactor variable name

kyucumber commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,24 +35,11 @@ public Question findQuestionById(Long id) {
+        Answers answers = new Answers(answerRepository.findByQuestionIdAndDeletedFalse(questionId));
+        DeleteHistories deleteHistories = new DeleteHistories();
+        deleteHistories.add(question.delete(loginUser));
+        deleteHistories.addAll(answers.delete(loginUser));
```
```
Question에서 answer delete까지 처리해주는 것이 맞을지 고민하였습니다. 하지만 양방향으로 맵핑하더라도 Qustion에서는 Answer를 조회만 할 수 있기 때문에 두가지 Question, Answer delete를 구분하여 Service에서 진행하였습니다. 이부분에 대해 이렇게 진행하여도 괜찮은지 의견 부탁드립니다.
```

지금과 같이 구현하셔도 큰 상관은 없어보입니다. 다만 양방향 연관관계 설정시 `cascade`와 같은 영속성 전이 설정을 이용해 `Question` 삭제 시 `Answers`가 지워지도록 구현할 수 있을 것 같네요.

kyucumber commented on src/test/java/qna/domain/QuestionRepositoryTest.java:
```diff
@@ -80,12 +86,49 @@ void findByDeletedFalseTest() {
+
+    @Test
+    @DisplayName("Answer list 조회 테스트")
+    void getAnswerList() {
+        Question savedQuestion = questionRepository.save(question1);
+        questionRepository.save(question2);
+
+        User answerWriter = new User("awriter", "password", "name", "sunju@slipp.net");
+        userRepository.save(answerWriter);
+        Answer answer = new Answer(answerWriter, savedQuestion, "Answers Contents");
+        answerRepository.save(answer);
+
+        questionRepository.flush();
+        assertThat(savedQuestion.getAnswers().size()).isEqualTo(1);
+    }
```
```
Question 양방향 맵핑 테스트를 해보려고 getAnswerList 테스트 코드를 작성하였으나 실패하였습니다. Question의 Answer list가 항상 길이가 0으로 나오고 있습니다. 혹시 어느부분이 문제인지 알 수 있을까요?
```

save를 통해 저장된 savedQuestion의 객체에는 이후에 저장된 answer에 대한 정보가 들어있지 않을 것 같네요.

기존 문제가 있던 코드는 아래와 같습니다.

```java
@Test
@DisplayName("Answer list 조회 테스트")
void getAnswerList() {
    Question savedQuestion = questionRepository.save(question1);
    questionRepository.save(question2);

    User answerWriter = new User("awriter", "password", "name", "sunju@slipp.net");
    userRepository.save(answerWriter);
    Answer answer = new Answer(answerWriter, savedQuestion, "Answers Contents");
    answerRepository.save(answer);

    questionRepository.flush();
    assertThat(savedQuestion.getAnswers().size()).isEqualTo(1);
}
```

여기서는 아래 두가지 방법을 통해 테스트를 수행할 수 있을것 같습니다.

1. Question의 addAnswer 메소드를 활용해 Question에도 Answer의 추가된 내용을 반영한다.

```java
@Test
@DisplayName("Answer list 조회 테스트")
void getAnswerList() {
    Question savedQuestion = questionRepository.save(question1);

    User answerWriter = new User("awriter", "password", "name", "sunju@slipp.net");
    userRepository.save(answerWriter);
    Answer answer = new Answer(answerWriter, savedQuestion, "Answers Contents");
    savedQuestion.addAnswer(answer);

    answerRepository.save(answer);

    assertThat(savedQuestion.getAnswers().size()).isEqualTo(1);
}
```

`Question`의 `addAnswer` 메소드도 다시 점검해보시는게 좋을 듯 합니다.. 제대로 구현이 되어있지 않네요

```java
public void addAnswer(Answer answer) {
    answers.add(answer); // 이 부분이 빠졌습니다.
    answer.toQuestion(this);
}
```

2. 영속성 컨텍스트를 비우고, 다시 조회 후 검증한다.

```java
@Test
@DisplayName("Answer list 조회 테스트")
void getAnswerList() {
    Question savedQuestion = questionRepository.save(question1);

    User answerWriter = new User("awriter", "password", "name", "sunju@slipp.net");
    userRepository.save(answerWriter);
    Answer answer = new Answer(answerWriter, savedQuestion, "Answers Contents");

    answerRepository.save(answer);

    entityManager.clear();

    Question updatedQuestion = questionRepository.findById(1L).get();
    assertThat(updatedQuestion.getAnswers().size()).isEqualTo(1);
}
```

jung635 replied:
아! 제가 잘 이해하고있지 못했던 부분이 있었네요. 두가지 다 테스트해보면서 알 수 있었습니다. 감사합니다!

kyucumber commented on src/main/java/qna/domain/Question.java:
```diff
@@ -66,8 +73,17 @@ public User getWriter() {
+    }
+
+    public List<Answer> getAnswers() {
+        return answers;
```
외부에서 값을 추가하거나 수정할 우려가 있어보이네요.

`Collections.unmodifibleList` 등을 통해 감싸서 내려주는건 어떨까요?

jung635 replied:
수정하였습니다 :)

kyucumber commented on src/main/java/qna/domain/Question.java:
```diff
@@ -66,8 +73,17 @@ public User getWriter() {
     }
 
-    public void delete() {
+    public DeleteHistory delete(User loginUser) throws CannotDeleteException {
+        if (!isOwner(loginUser)) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
         this.deleted = true;
+
+        return new DeleteHistory(ContentType.QUESTION, this, LocalDateTime.now());
+    }
```
`Question`에서 삭제 후 `DeleteHistory`를 반환하고 있네요.

`Question`, `Answer` 등 여러 객체에서 `DeleteHistory`와 결합을 가지는 경우 `DeleteHistory`의 변경에 `Question`, `Answer`가 영향을 받을 우려가 있어 보입니다.

`DeleteHistory`는 늘어나는 모든 엔티티에서 사용할 가능성이 높아 보이는데 모든 엔티티에서 결합을 가지도록 설계하는게 좋은 방향일지 고민해보세요.

jung635 replied:
question을 받아서 deleteHistory list를 생성하도록 하였습니다. 혹시 더 나은 방향이 있으면 말씀 부탁드립니다!

kyucumber commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -75,8 +77,13 @@ public User getWriter() {
 
-    public void delete(boolean deleted) {
-        this.deleted = deleted;
+    public DeleteHistory delete(User loginUser) throws CannotDeleteException {
+        if (!isOwner(loginUser)) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+        }
+        this.deleted = true;
+
+        return new DeleteHistory(ContentType.ANSWER, this, LocalDateTime.now());
     }
```
아래 `Question`에 현재 구현과 관련해 코멘트 남겨두었으니 확인 부탁드립니다.

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

jung635 committed:
fix($step3): fix test cases for Question

jung635 committed:
feat($step3): change Question and QnaService

add Answers embeded to Question
split to create deleteHistories logic in QnaService

jung635 committed:
refactor($step3): refactor QnaService

change to add deleteHistories logic

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지할게요
